### PR TITLE
Unique point matrices in `NCMesh::NCList` [nc-point-mat-index-dev]

### PR DESCRIPTION
**Pull request by jakubcerveny**
Currently, each `NCMesh::Slave` holds a  `DenseMatrix point_matrix`. These matrices often repeat, since typically many coarse-fine situations are geometrically similar.

This PR replaces the per-slave matrix with an index into an array of unique point matrices (`NCMesh::NCList::point_matrices`). This has two advantages:
1. Some memory is saved, for example when refining `inline-hex.mesh` randomly 6 times, one obtains a mesh with ~429100 elements, and the RAM taken by `NCMesh::face_list` and `NCMesh::edge_list` goes from 242 MB to 91MB (this is rather extreme due to the randomness).
2. More importantly, slave faces/edges are now basically split into classes based on their position and orientation within the master entity. This can be useful when batching the assembly of DG fluxes (cc: @YohannDudouit)

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1749](https://github.com/mfem/mfem/pull/1749) | @jakubcerveny | @v-dobrev | @YohannDudouit + @rw-anderson | 09/10/20 | 10/10/20 | 10/21/20 | |

### CODE REVIEW

**Comment by jakubcerveny** [comment]
@YohannDudouit please let me know if this is sufficient for you. One thing that may be missing here is that `Mesh::NCFaceInfo` doesn't specify the matrix index and at the same time some of the faces get a differently oriented matrix (see the end of `ParNCMesh::GetFaceNeighbors`).

File: mesh/ncmesh.cpp [hunk 1]
```diff
@@ -2311,8 +2312,61 @@ int NCMesh::find_local_face(int geom, int a, int b, int c)
...
+      {
+         for (int j = 0; j < pm.points[i].dim; j++)
+         {
+            double coord = pm.points[i].coord[j];
+            hash = 31*hash + *((std::uint64_t*) &coord);
+         }
+      }
+      return hash;
+   }
+};
```
Comment (YohannDudouit) [inline mesh/ncmesh.cpp:2332]: Why is this a good hash key?

File: mesh/ncmesh.cpp [hunk 2]
```diff
@@ -3398,6 +3472,20 @@ void NCMesh::FindVertexCousins(int elem, int local, Array<int> &cousins) const
 
 //// Coarse/fine transformations ///////////////////////////////////////////////
 
+bool NCMesh::PointMatrix::operator==(const PointMatrix &pm) const
+{
+   MFEM_ASSERT(np == pm.np, "");
```
Comment (YohannDudouit) [inline mesh/ncmesh.cpp:3477]: Shouldn't this return `false` instead of returning an error?

File: mesh/ncmesh.cpp [hunk 3]
```diff
@@ -3398,6 +3472,20 @@ void NCMesh::FindVertexCousins(int elem, int local, Array<int> &cousins) const
...
+   for (int i = 0; i < np; i++)
+   {
+      MFEM_ASSERT(points[i].dim == pm.points[i].dim, "");
```
Comment (YohannDudouit) [inline mesh/ncmesh.cpp:3480]: same here.

**Review by YohannDudouit — COMMENTED** [review]
Thank you @jakubcerveny for doing this, overall this is almost good to be merged, I would just like to have some documentation to make the intent/purpose limpid.

File: mesh/ncmesh.cpp [hunk 4]
```diff
@@ -2311,8 +2312,61 @@ int NCMesh::find_local_face(int geom, int a, int b, int c)
...
+   {
+      for (unsigned i = 0; i < map.bucket_count(); i++)
+      {
+         mfem::out << map.bucket_size(i) << " ";
+      }
+   }
+
+private:
+   std::unordered_map<NCMesh::PointMatrix, int, PointMatrixHash> map;
+};
```
[the remark this answers is not in the document]
Comment (jakubcerveny) [reply mesh/ncmesh.cpp:2367]: Added docstring for MatrixMap.

File: mesh/ncmesh.cpp [hunk 5]
```diff
@@ -2792,70 +2865,71 @@ void NCMesh::BuildVertexList()
...
+   oriented_matrix = *(point_matrices[slave.Geom()][slave.matrix]);
 
-   if (edge_flags)
+   if (slave.edge_flags)
    {
       MFEM_ASSERT(oriented_matrix.Height() == 1 &&
                   oriented_matrix.Width() == 2, "not an edge point matrix");
 
-      if (edge_flags & 1) // master inverted
+      if (slave.edge_flags & 1) // master inverted
```
[the remark this answers is not in the document]
Comment (jakubcerveny) [reply mesh/ncmesh.cpp:2887]: The point matrix is needed 1) to get the master/slave interpolation weights in FESpace and ParFESpace and 2) to construct the ParMesh face neighbors (in ParNCMesh::GetFaceNeighbors). These two handle face/edge orientation differently, so the point matrix is stored in a "neutral" format (at least that was the intent IIRC) and then OrientedPointMatrix fixes the edge orientation if necessary for case 1), if the flag is set. Getting the orientations right in all cases was always a bit of a mess and honestly I don't know how the master can be inverted, from the code this flag seems to be no longer used.

File: mesh/ncmesh.cpp [hunk 6]
```diff
@@ -2792,70 +2865,71 @@ void NCMesh::BuildVertexList()
...
       {
          oriented_matrix(0,0) = 1.0 - oriented_matrix(0,0);
          oriented_matrix(0,1) = 1.0 - oriented_matrix(0,1);
       }
-      if (edge_flags & 2) // slave inverted
+      if (slave.edge_flags & 2) // slave inverted
```
[the remark this answers is not in the document]
Comment (jakubcerveny) [reply mesh/ncmesh.cpp:2892]: This is definitely still necessary, so each slave edge has this bit flag that tells OrientedPointMatrix to invert the orientation.

File: mesh/ncmesh.hpp [hunk 7]
```diff
@@ -178,32 +179,37 @@ class NCMesh
    struct Slave : public MeshId
    {
       int master; ///< master number (in Mesh numbering)
-      int edge_flags; ///< edge orientation flags
-      DenseMatrix point_matrix; ///< position within the master edge/face
+      unsigned matrix : 24;    ///< index into NCList::point_matrices[geom]
+      unsigned edge_flags : 8; ///< edge orientation flags
```
[the remark this answers is not in the document]
Comment (jakubcerveny) [reply mesh/ncmesh.hpp:183]: I didn't want this to take 8 bytes, even the 24+8 bits is more than enough. Hopefully the struct is packed as if this was just one int. The matrix index obviously needs more bits than the flags, but 24 is enough.

File: mesh/ncmesh.cpp (hunk 1 above)
Comment (jakubcerveny) [reply mesh/ncmesh.cpp:2332]: This a variation on the section "Hashing an array of floats" from here: https://cs.stackexchange.com/questions/37952/hash-function-floating-point-inputs-for-genetic-algorithm

I tested this simple version and the original that had more mixing of the bits at the end, but the distribution into the buckets of unordered_map was practically unaffected. The important thing to check is that there aren't any unfortunate collisions (some buckets overfilled), I did that using "DumpBucketSizes" which is left in the struct for potential debugging. 

The hash seems to work OK for our purpose. The initial constant is more or less random (only important if the matrix was empty). The constant 31 is more important, I think it's good that it's a prime. Beyond this and some basic testing I didn't put more thought into this.

File: mesh/ncmesh.cpp (hunk 2 above)
Comment (jakubcerveny) [reply mesh/ncmesh.cpp:3480]: No, in this case comparing points with incompatible dimensions is an internal error.

File: mesh/ncmesh.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jakubcerveny) [reply mesh/ncmesh.cpp:2887]: In any case this is unrelated to this PR...

File: mesh/ncmesh.cpp (hunk 1 above)
Comment (pazner) [reply mesh/ncmesh.cpp:2332]: Some comments or documentation could help a lot here, especially given the "magic number" literals.

Comment (jakubcerveny) [reply mesh/ncmesh.cpp]: Added more comments for the hash functor.

File: mesh/ncmesh.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply mesh/ncmesh.cpp:2898]: I wasn't really asking if it's necessary or not, I was wondering how I'm supposed to know that the magic number `2` means "slave inverted".

**Review by YohannDudouit — APPROVED** [review]
Thank you for doing this so quickly @jakubcerveny !

**Comment by tzanio** [comment]
@rw-anderson can you please review?

File: mesh/ncmesh.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (rw-anderson) [reply mesh/ncmesh.hpp:183]: Is there a list of what the edge flags are somewhere?

**Comment by rw-anderson** [comment]
I left one more question about docs for edge_flags.  Other than that looks good.

**Comment by jakubcerveny** [comment]
@v-dobrev, this should be good to go.

**Comment by jakubcerveny** [comment]
@v-dobrev, this is good to go into `next`...

**Review by rw-anderson — APPROVED** [review]

**Comment by v-dobrev** [comment]
LGTM

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
The internal autotests showed that the following two sample runs fail:
```console
ex15 -m ../data/escher.mesh -r 2 -tf 0.3 -no-vis
mpirun -np 4 ex15p -m ../data/escher.mesh -r 2 -tf 0.3 -no-vis
```
Running in debug mode triggers an assert failure which should help with tracking down the issue.

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Pull request merged**